File: ios/BetterW4/OnDutyModels.swift
//
//  OnDutyModels.swift
//  BetterW4
//
//  Who is on duty today — `people/onduty` plus the month calendar at
//  `people/onduty/schedule` (features.md §1.12).
//
//  Evidence: live capture 19 Aug 2026. Today's page prints one card per
//  person under an `<h3>` role (`House Leader on Call`), with photo, name,
//  Phone, E-mail and Location. The calendar prints `.onduty-group-name` +
//  `.onduty-group` inside `td.day`.
//

import Foundation

struct OnDutyPerson: Identifiable, Hashable, Sendable {
    let id: String
    let name: String
    let role: String
    let uwcId: String?
    let phone: String?
    let email: String?
    let location: String?
    let photoURL: URL?

    var hasContact: Bool {
        !(phone ?? "").isEmpty || !(email ?? "").isEmpty
    }
}

struct OnDutyGroup: Identifiable, Hashable, Sendable {
    var id: String { role }
    let role: String
    let people: [OnDutyPerson]
}

struct OnDutyPage: Hashable, Sendable {
    var title: String?
    var date: Date?
    var dateLabel: String?
    var groups: [OnDutyGroup]

    var people: [OnDutyPerson] { groups.flatMap(\.people) }
    var isEmpty: Bool { people.isEmpty }

    init(
        title: String? = nil,
        date: Date? = nil,
        dateLabel: String? = nil,
        groups: [OnDutyGroup] = []
    ) {
        self.title = title
        self.date = date
        self.dateLabel = dateLabel
        self.groups = groups
    }
}

struct OnDutyDay: Identifiable, Hashable, Sendable {
    let id: String
    let date: Date?
    let dateLabel: String
    let isToday: Bool
    let groups: [OnDutyGroup]

    var people: [OnDutyPerson] { groups.flatMap(\.people) }
}

struct OnDutySchedule: Hashable, Sendable {
    var monthLabel: String?
    var year: Int?
    var month: Int?
    var days: [OnDutyDay]

    init(
        monthLabel: String? = nil,
        year: Int? = nil,
        month: Int? = nil,
        days: [OnDutyDay] = []
    ) {
        self.monthLabel = monthLabel
        self.year = year
        self.month = month
        self.days = days
    }
}

struct OnDutySnapshot: Hashable, Sendable {
    var today: OnDutyPage
    var upcoming: [OnDutyDay]
}

enum OnDutyContact {
    static func telephoneURL(_ phone: String) -> URL? {
        guard let digits = digitsForDialing(phone) else { return nil }
        return URL(string: "tel:\(digits)")
    }

    static func smsURL(_ phone: String) -> URL? {
        guard let digits = digitsForDialing(phone) else { return nil }
        return URL(string: "sms:\(digits)")
    }

    static func mailtoURL(_ email: String) -> URL? {
        let trimmed = email.trimmingCharacters(in: .whitespacesAndNewlines)
        guard trimmed.contains("@"), !trimmed.isEmpty else { return nil }
        let encoded = trimmed.addingPercentEncoding(withAllowedCharacters: .urlPathAllowed) ?? trimmed
        return URL(string: "mailto:\(encoded)")
    }

    static func digitsForDialing(_ phone: String) -> String? {
        var digits = ""
        for character in phone {
            if character.isNumber {
                digits.append(character)
            } else if character == "+", digits.isEmpty {
                digits.append(character)
            }
        }
        let digitCount = digits.filter(\.isNumber).count
        return digitCount >= 5 ? digits : nil
    }
}

File: android/app/src/main/java/dk/betterw4/android/feature/onduty/OnDutyModels.kt
package dk.betterw4.android.feature.onduty

import android.net.Uri
import java.time.LocalDate

data class OnDutyPerson(
    val id: String,
    val name: String,
    val role: String,
    val uwcId: String? = null,
    val phone: String? = null,
    val email: String? = null,
    val location: String? = null,
    val photoUrl: String? = null,
) {
    val hasContact: Boolean get() = !phone.isNullOrBlank() || !email.isNullOrBlank()
}

data class OnDutyGroup(
    val role: String,
    val people: List<OnDutyPerson>,
)

data class OnDutyPage(
    val title: String? = null,
    val date: LocalDate? = null,
    val dateLabel: String? = null,
    val groups: List<OnDutyGroup> = emptyList(),
) {
    val people: List<OnDutyPerson> get() = groups.flatMap { it.people }
    val isEmpty: Boolean get() = people.isEmpty()
}

data class OnDutyDay(
    val id: String,
    val date: LocalDate?,
    val dateLabel: String,
    val isToday: Boolean = false,
    val groups: List<OnDutyGroup> = emptyList(),
) {
    val people: List<OnDutyPerson> get() = groups.flatMap { it.people }
}

data class OnDutySchedule(
    val monthLabel: String? = null,
    val year: Int? = null,
    val month: Int? = null,
    val days: List<OnDutyDay> = emptyList(),
)

data class OnDutySnapshot(
    val today: OnDutyPage,
    val upcoming: List<OnDutyDay> = emptyList(),
)

object OnDutyContact {
    fun telephoneUri(phone: String): Uri? {
        val digits = digitsForDialing(phone) ?: return null
        return Uri.parse("tel:$digits")
    }

    fun smsUri(phone: String): Uri? {
        val digits = digitsForDialing(phone) ?: return null
        return Uri.parse("smsto:$digits")
    }

    fun mailtoUri(email: String): Uri? {
        val trimmed = email.trim()
        if (trimmed.isEmpty() || !trimmed.contains('@')) return null
        return Uri.parse("mailto:${Uri.encode(trimmed)}")
    }

    fun digitsForDialing(phone: String): String? {
        val builder = StringBuilder()
        for (ch in phone) {
            when {
                ch.isDigit() -> builder.append(ch)
                ch == '+' && builder.isEmpty() -> builder.append(ch)
            }
        }
        val digits = builder.toString()
        val digitCount = digits.count { it.isDigit() }
        return digits.takeIf { digitCount >= 5 }
    }
}
